src/utils/object_assign.ts

```diff
@@ -14,25 +14,58 @@
  * limitations under the License.
  */
 
+
+type MergeRecursively<T extends object, S extends object[]> =
+    S extends [infer First, ...infer Rest] ?
+      MergeRecursively<MergeObjects<T, First extends object ? First : never>,
+       Rest extends object[] ? Rest: never>
+    : T;
+;
+
+type MergeObjects<T extends object, S extends object> = {
+  [K in (OptionalPropertyOfTNotInS<T, S> | OptionalProperty<T>)]?:
+        K extends keyof S ? S[K]
+        : K extends keyof T ? T[K]
+        : never;
+} & {
+  [K in Exclude<
+    (keyof T | keyof S),
+    OptionalPropertyOfTNotInS<T, S> | OptionalProperty<T>
+    >]:
+        K extends keyof S ? S[K]
+        : K extends keyof T ? T[K]
+        : never;
+};
+
+type OptionalProperty<T> = Exclude<{
+  [K in keyof T]: T extends Record<K, T[K]>
+  ? never
+  : K
+}[keyof T], undefined>;
+
+type OptionalPropertyOfTNotInS<T, S> = Exclude<{
+  [K in keyof T]: T extends Record<K, T[K]>
+      ? never
+      : K
+}[keyof T], undefined | {[K in keyof S]: S extends Record<K, S[K]>
+  ? K
+  : never
+}[keyof S]>;
```

Comment:
Where did you find that trick?

How is Object.assign typed with TypeScript on that matter?

Reply:
I find this article quite useful, even if the goal is a bit different because they are doing deep merging
https://dev.to/svehla/typescript-how-to-deep-merge-170c

Object.assign typing is defined here and is equal to the intersection of the input types:
https://github.com/microsoft/TypeScript/blob/main/src/lib/es2015.core.d.ts#L259